File: gospider/media/downloader.go
package media

import (
	"encoding/json"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"regexp"
)

// MediaDownloader 媒体下载器
type MediaDownloader struct {
	client      *http.Client
	outputDir   string
	userAgent   string
}

// NewMediaDownloader 创建媒体下载器
func NewMediaDownloader(outputDir string) *MediaDownloader {
	return &MediaDownloader{
		client:    &http.Client{},
		outputDir: outputDir,
		userAgent: "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
	}
}

// DownloadResult 下载结果
type DownloadResult struct {
	URL      string `json:"url"`
	Path     string `json:"path"`
	Size     int64  `json:"size"`
	Error    string `json:"error,omitempty"`
	Success  bool   `json:"success"`
}

// DownloadImage 下载图片
func (md *MediaDownloader) DownloadImage(url, filename string) *DownloadResult {
	result := &DownloadResult{URL: url}
	
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		result.Error = err.Error()
		result.Success = false
		return result
	}
	
	req.Header.Set("User-Agent", md.userAgent)
	req.Header.Set("Referer", "https://www.google.com/")
	
	resp, err := md.client.Do(req)
	if err != nil {
		result.Error = err.Error()
		result.Success = false
		return result
	}
	defer resp.Body.Close()
	
	if resp.StatusCode != 200 {
		result.Error = "HTTP " + resp.Status
		result.Success = false
		return result
	}
	
	// 创建输出目录
	os.MkdirAll(md.outputDir, 0755)
	
	// 生成文件名
	if filename == "" {
		filename = filepath.Base(url)
		if filename == "" {
			filename = "image.jpg"
		}
	}
	
	filepath := filepath.Join(md.outputDir, filename)
	
	// 保存文件
	file, err := os.Create(filepath)
	if err != nil {
		result.Error = err.Error()
		result.Success = false
		return result
	}
	defer file.Close()
	
	size, err := io.Copy(file, resp.Body)
	if err != nil {
		result.Error = err.Error()
		result.Success = false
		return result
	}
	
	result.Path = filepath
	result.Size = size
	result.Success = true
	
	return result
}

// DownloadImages 批量下载图片
func (md *MediaDownloader) DownloadImages(urls []string) []*DownloadResult {
	results := make([]*DownloadResult, 0, len(urls))
	
	for _, url := range urls {
		result := md.DownloadImage(url, "")
		results = append(results, result)
	}
	
	return results
}

// ExtractImagesFromHTML 从 HTML 中提取图片链接
func (md *MediaDownloader) ExtractImagesFromHTML(html string) []string {
	var images []string
	
	// 匹配 img 标签的 src 属性
	re := regexp.MustCompile(`<img[^>]+src=["']([^"']+)["']`)
	matches := re.FindAllStringSubmatch(html, -1)
	
	for _, match := range matches {
		if len(match) > 1 {
			images = append(images, match[1])
		}
	}
	
	return images
}

// DownloadVideo 下载视频
func (md *MediaDownloader) DownloadVideo(url, filename string) *DownloadResult {
	result := &DownloadResult{URL: url}
	
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		result.Error = err.Error()
		result.Success = false
		return result
	}
	
	req.Header.Set("User-Agent", md.userAgent)
	req.Header.Set("Range", "bytes=0-")
	
	resp, err := md.client.Do(req)
	if err != nil {
		result.Error = err.Error()
		result.Success = false
		return result
	}
	defer resp.Body.Close()
	
	if resp.StatusCode != 200 && resp.StatusCode != 206 {
		result.Error = "HTTP " + resp.Status
		result.Success = false
		return result
	}
	
	// 创建输出目录
	videoDir := filepath.Join(md.outputDir, "videos")
	os.MkdirAll(videoDir, 0755)
	
	// 生成文件名
	if filename == "" {
		filename = "video.mp4"
	}
	
	filepath := filepath.Join(videoDir, filename)
	
	// 保存文件
	file, err := os.Create(filepath)
	if err != nil {
		result.Error = err.Error()
		result.Success = false
		return result
	}
	defer file.Close()
	
	size, err := io.Copy(file, resp.Body)
	if err != nil {
		result.Error = err.Error()
		result.Success = false
		return result
	}
	
	result.Path = filepath
	result.Size = size
	result.Success = true
	
	return result
}

// DownloadAudio 下载音频
func (md *MediaDownloader) DownloadAudio(url, filename string) *DownloadResult {
	result := &DownloadResult{URL: url}
	
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		result.Error = err.Error()
		result.Success = false
		return result
	}
	
	req.Header.Set("User-Agent", md.userAgent)
	
	resp, err := md.client.Do(req)
	if err != nil {
		result.Error = err.Error()
		result.Success = false
		return result
	}
	defer resp.Body.Close()
	
	if resp.StatusCode != 200 {
		result.Error = "HTTP " + resp.Status
		result.Success = false
		return result
	}
	
	// 创建输出目录
	audioDir := filepath.Join(md.outputDir, "audio")
	os.MkdirAll(audioDir, 0755)
	
	// 生成文件名
	if filename == "" {
		filename = "audio.mp3"
	}
	
	filepath := filepath.Join(audioDir, filename)
	
	// 保存文件
	file, err := os.Create(filepath)
	if err != nil {
		result.Error = err.Error()
		result.Success = false
		return result
	}
	defer file.Close()
	
	size, err := io.Copy(file, resp.Body)
	if err != nil {
		result.Error = err.Error()
		result.Success = false
		return result
	}
	
	result.Path = filepath
	result.Size = size
	result.Success = true
	
	return result
}

// VideoInfo 视频信息
type VideoInfo struct {
	Title       string `json:"title"`
	Duration    int    `json:"duration"`
	Thumbnail   string `json:"thumbnail"`
	Formats     []VideoFormat `json:"formats"`
}

// VideoFormat 视频格式
type VideoFormat struct {
	FormatID  string `json:"format_id"`
	Ext       string `json:"ext"`
	Resolution string `json:"resolution"`
	Filesize  int64  `json:"filesize"`
	URL       string `json:"url"`
}

// GetVideoInfo 获取视频信息（模拟实现）
func (md *MediaDownloader) GetVideoInfo(url string) (*VideoInfo, error) {
	// 实际应该调用 yt-dlp 或类似工具
	// 这里只是示例
	return &VideoInfo{
		Title: "Video Title",
	}, nil
}

// ExtractMediaURLs 从页面提取媒体 URL
func (md *MediaDownloader) ExtractMediaURLs(html string) MediaURLs {
	urls := MediaURLs{
		Images: make([]string, 0),
		Videos: make([]string, 0),
		Audios: make([]string, 0),
	}
	
	// 提取图片
	imgRe := regexp.MustCompile(`(https?://[^\s"'<>]+\.(?:jpg|jpeg|png|gif|webp))`)
	urls.Images = appendUnique(urls.Images, imgRe.FindAllString(html, -1)...)
	
	// 提取视频
	videoRe := regexp.MustCompile(`(https?://[^\s"'<>]+\.(?:mp4|webm|avi|mov|flv))`)
	urls.Videos = appendUnique(urls.Videos, videoRe.FindAllString(html, -1)...)
	
	// 提取音频
	audioRe := regexp.MustCompile(`(https?://[^\s"'<>]+\.(?:mp3|wav|ogg|flac|aac))`)
	urls.Audios = appendUnique(urls.Audios, audioRe.FindAllString(html, -1)...)
	
	return urls
}

// MediaURLs 媒体 URL 集合
type MediaURLs struct {
	Images []string `json:"images"`
	Videos []string `json:"videos"`
	Audios []string `json:"audios"`
}

func appendUnique(slice []string, items ...string) []string {
	for _, item := range items {
		found := false
		for _, existing := range slice {
			if existing == item {
				found = true
				break
			}
		}
		if !found {
			slice = append(slice, item)
		}
	}
	return slice
}

// DownloadAll 下载所有媒体
func (md *MediaDownloader) DownloadAll(urls MediaURLs) DownloadStats {
	stats := DownloadStats{}
	
	// 下载图片
	for _, url := range urls.Images {
		result := md.DownloadImage(url, "")
		if result.Success {
			stats.ImagesDownloaded++
			stats.TotalBytes += result.Size
		} else {
			stats.ImagesFailed++
		}
	}
	
	// 下载视频
	for _, url := range urls.Videos {
		result := md.DownloadVideo(url, "")
		if result.Success {
			stats.VideosDownloaded++
			stats.TotalBytes += result.Size
		} else {
			stats.VideosFailed++
		}
	}
	
	// 下载音频
	for _, url := range urls.Audios {
		result := md.DownloadAudio(url, "")
		if result.Success {
			stats.AudiosDownloaded++
			stats.TotalBytes += result.Size
		} else {
			stats.AudiosFailed++
		}
	}
	
	return stats
}

// DownloadStats 下载统计
type DownloadStats struct {
	ImagesDownloaded  int   `json:"images_downloaded"`
	ImagesFailed      int   `json:"images_failed"`
	VideosDownloaded  int   `json:"videos_downloaded"`
	VideosFailed      int   `json:"videos_failed"`
	AudiosDownloaded  int   `json:"audios_downloaded"`
	AudiosFailed      int   `json:"audios_failed"`
	TotalBytes        int64 `json:"total_bytes"`
}

func (ds *DownloadStats) String() string {
	data, _ := json.MarshalIndent(ds, "", "  ")
	return string(data)
}

File: pyspider/media/downloader.py
"""
媒体下载模块
支持图片、视频、音乐爬取
"""

import os
import re
import json
import requests
from typing import List, Dict, Optional
from dataclasses import dataclass
from pathlib import Path


@dataclass
class DownloadResult:
    """下载结果"""
    url: str
    path: str
    size: int
    success: bool
    error: Optional[str] = None


@dataclass
class VideoInfo:
    """视频信息"""
    title: str
    duration: int
    thumbnail: str
    formats: List[Dict]


class MediaDownloader:
    """媒体下载器"""
    
    def __init__(self, output_dir: str = "./downloads"):
        self.output_dir = Path(output_dir)
        self.session = requests.Session()
        self.session.headers.update({
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
        })
        self._create_dirs()
    
    def _create_dirs(self) -> None:
        """创建目录"""
        (self.output_dir / "images").mkdir(parents=True, exist_ok=True)
        (self.output_dir / "videos").mkdir(parents=True, exist_ok=True)
        (self.output_dir / "audio").mkdir(parents=True, exist_ok=True)
    
    def download_image(
        self,
        url: str,
        filename: str = None,
        referer: str = None,
    ) -> DownloadResult:
        """下载图片"""
        try:
            headers = {}
            if referer:
                headers["Referer"] = referer
            
            resp = self.session.get(url, headers=headers, timeout=30)
            resp.raise_for_status()
            
            # 生成文件名
            if not filename:
                filename = os.path.basename(url.split("?")[0])
                if not filename:
                    filename = f"image_{len(os.listdir(self.output_dir / 'images'))}.jpg"
            
            filepath = self.output_dir / "images" / filename
            
            # 保存文件
            with open(filepath, "wb") as f:
                f.write(resp.content)
            
            return DownloadResult(
                url=url,
                path=str(filepath),
                size=len(resp.content),
                success=True,
            )
        except Exception as e:
            return DownloadResult(
                url=url,
                path="",
                size=0,
                success=False,
                error=str(e),
            )
    
    def download_images(self, urls: List[str]) -> List[DownloadResult]:
        """批量下载图片"""
        results = []
        for url in urls:
            result = self.download_image(url)
            results.append(result)
        return results
    
    def download_video(
        self,
        url: str,
        filename: str = None,
        chunk_size: int = 8192,
    ) -> DownloadResult:
        """下载视频"""
        try:
            resp = self.session.get(url, stream=True, timeout=30)
            resp.raise_for_status()
            
            # 生成文件名
            if not filename:
                filename = f"video_{len(os.listdir(self.output_dir / 'videos'))}.mp4"
            
            filepath = self.output_dir / "videos" / filename
            
            # 保存文件
            with open(filepath, "wb") as f:
                for chunk in resp.iter_content(chunk_size=chunk_size):
                    if chunk:
                        f.write(chunk)
            
            return DownloadResult(
                url=url,
                path=str(filepath),
                size=filepath.stat().st_size,
                success=True,
            )
        except Exception as e:
            return DownloadResult(
                url=url,
                path="",
                size=0,
                success=False,
                error=str(e),
            )
    
    def download_audio(
        self,
        url: str,
        filename: str = None,
        chunk_size: int = 8192,
    ) -> DownloadResult:
        """下载音频"""
        try:
            resp = self.session.get(url, stream=True, timeout=30)
            resp.raise_for_status()
            
            # 生成文件名
            if not filename:
                filename = f"audio_{len(os.listdir(self.output_dir / 'audio'))}.mp3"
            
            filepath = self.output_dir / "audio" / filename
            
            # 保存文件
            with open(filepath, "wb") as f:
                for chunk in resp.iter_content(chunk_size=chunk_size):
                    if chunk:
                        f.write(chunk)
            
            return DownloadResult(
                url=url,
                path=str(filepath),
                size=filepath.stat().st_size,
                success=True,
            )
        except Exception as e:
            return DownloadResult(
                url=url,
                path="",
                size=0,
                success=False,
                error=str(e),
            )
    
    def extract_images_from_html(self, html: str) -> List[str]:
        """从 HTML 中提取图片链接"""
        urls = []
        
        # 匹配 img 标签
        img_pattern = r'<img[^>]+src=["\']([^"\']+)["\']'
        matches = re.findall(img_pattern, html, re.IGNORECASE)
        urls.extend(matches)
        
        # 匹配背景图片
        bg_pattern = r'url\(["\']?([^"\')\s]+)["\']?\)'
        bg_matches = re.findall(bg_pattern, html, re.IGNORECASE)
        urls.extend([url for url in bg_matches if any(ext in url for ext in [".jpg", ".jpeg", ".png", ".gif", ".webp"])])
        
        # 去重和过滤
        urls = list(set(urls))
        urls = [url for url in urls if url.startswith(("http://", "https://"))]
        
        return urls
    
    def extract_media_urls(self, html: str) -> "MediaURLs":
        """从 HTML 中提取所有媒体 URL"""
        urls = MediaURLs()
        
        # 提取图片
        img_pattern = r'(https?://[^\s"\'\'>]+\.(?:jpg|jpeg|png|gif|webp|bmp))'
        urls.images = list(set(re.findall(img_pattern, html, re.IGNORECASE)))
        
        # 提取视频
        video_pattern = r'(https?://[^\s"\'\'>]+\.(?:mp4|webm|avi|mov|flv|mkv))'
        urls.videos = list(set(re.findall(video_pattern, html, re.IGNORECASE)))
        
        # 提取音频
        audio_pattern = r'(https?://[^\s"\'\'>]+\.(?:mp3|wav|ogg|flac|aac|m4a))'
        urls.audios = list(set(re.findall(audio_pattern, html, re.IGNORECASE)))
        
        return urls
    
    def download_all(self, media_urls: "MediaURLs") -> "DownloadStats":
        """下载所有媒体"""
        stats = DownloadStats()
        
        # 下载图片
        for url in media_urls.images:
            result = self.download_image(url)
            if result.success:
                stats.images_downloaded += 1
                stats.total_bytes += result.size
            else:
                stats.images_failed += 1
        
        # 下载视频
        for url in media_urls.videos:
            result = self.download_video(url)
            if result.success:
                stats.videos_downloaded += 1
                stats.total_bytes += result.size
            else:
                stats.videos_failed += 1
        
        # 下载音频
        for url in media_urls.audios:
            result = self.download_audio(url)
            if result.success:
                stats.audios_downloaded += 1
                stats.total_bytes += result.size
            else:
                stats.audios_failed += 1
        
        return stats
    
    def get_video_info(self, url: str) -> Optional[VideoInfo]:
        """获取视频信息（模拟实现）"""
        # 实际应该调用 yt-dlp 或类似工具
        return VideoInfo(
            title="Video Title",
            duration=0,
            thumbnail="",
            formats=[],
        )


@dataclass
class MediaURLs:
    """媒体 URL 集合"""
    images: List[str] = None
    videos: List[str] = None
    audios: List[str] = None
    
    def __post_init__(self):
        if self.images is None:
            self.images = []
        if self.videos is None:
            self.videos = []
        if self.audios is None:
            self.audios = []


@dataclass
class DownloadStats:
    """下载统计"""
    images_downloaded: int = 0
    images_failed: int = 0
    videos_downloaded: int = 0
    videos_failed: int = 0
    audios_downloaded: int = 0
    audios_failed: int = 0
    total_bytes: int = 0
    
    def to_dict(self) -> Dict:
        """转换为字典"""
        return {
            "images_downloaded": self.images_downloaded,
            "images_failed": self.images_failed,
            "videos_downloaded": self.videos_downloaded,
            "videos_failed": self.videos_failed,
            "audios_downloaded": self.audios_downloaded,
            "audios_failed": self.audios_failed,
            "total_bytes": self.total_bytes,
        }
    
    def __str__(self) -> str:
        """字符串表示"""
        return json.dumps(self.to_dict(), indent=2, ensure_ascii=False)
